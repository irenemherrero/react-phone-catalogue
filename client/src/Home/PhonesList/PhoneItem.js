import React from 'react'
import PropTypes from 'prop-types'
import { AvailableColors, PhoneImage } from '../../common'
import { ResponsiveWrapper, PhoneCard, PhoneCardInfo } from './styles'

/**
 * Showing phone item
 * @component
 */
const PhoneItem = (props) => {
  const { id, name, manufacturer, color, price, imageFileName, onClick } = props
  const urlImage = process.env.PHONES_IMAGES_URL + imageFileName
  return (
    <ResponsiveWrapper onClick={() => onClick(id)}>
      <PhoneCard>
        <PhoneImage src={urlImage} alt={name} />
        <PhoneCardInfo>
          <h1>{name}</h1>
          <h2>{manufacturer}</h2>
          <AvailableColors colors={color.split(',')} />
          <footer>{`${price} â‚¬`}</footer>
        </PhoneCardInfo>
      </PhoneCard>
    </ResponsiveWrapper>
  )
}

PhoneItem.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  manufacturer: PropTypes.string,
  /**comma separated colors */
  color: PropTypes.string,
  price: PropTypes.number,
  imageFileName: PropTypes.string,
  /**fires parent onClick */
  onClick: PropTypes.func,
}
PhoneItem.defaultProps = {
  manufacturer: 'N/A',
  imageFileName: 'noimage.png',
  price: 0,
  onClick: () => {},
}

export { PhoneItem }
