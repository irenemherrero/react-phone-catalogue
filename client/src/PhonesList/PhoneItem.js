import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import tw from 'tailwind.macro'

const ResponsiveContainer = styled.div`
  ${tw`w-full md:w-1/2 lg:w-1/3 px-2 my-2`}
`
const PhoneCard = styled.div`
  ${tw`flex flex-row shadow-md bg-white cursor-pointer hover:shadow-xl`}
  &:hover {
    transform: scale(1.04);
  }
`
const PhoneCardImage = styled.img`
  ${tw`h-48 w-40`}
`
const PhoneCardInfo = styled.div`
  ${tw`flex flex-col p-6`}
  h1 {
    ${tw`font-semibold text-lg`}
  }
  p {
    ${tw`text-gray-600`}
  }
`
const ColorsContainer = styled.div`
  ${tw`flex flex-row mt-6`}
`

const PhoneColor = styled.span`
  ${tw`w-10 h-5 mr-1 border rounded`}
  background-color: ${(props) => props.color}
`

const PhoneItem = (props) => {
  const { id, name, onClick, imageFileName, price, color } = props
  const urlImage = process.env.PHONES_IMAGES_URL + imageFileName
  return (
    <ResponsiveContainer onClick={() => onClick(id)}>
      <PhoneCard>
        <div className="flex-shrink-0">
          <PhoneCardImage src={urlImage} alt={name} />
        </div>
        <PhoneCardInfo>
          <h1>{name}</h1>
          <p>{price}â‚¬</p>
          <ColorsContainer>
            {color.split(',').map((c) => (
              <PhoneColor key={c} color={c} />
            ))}
          </ColorsContainer>
        </PhoneCardInfo>
      </PhoneCard>
    </ResponsiveContainer>
  )
}

PhoneItem.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  imageFileName: PropTypes.string,
  price: PropTypes.number,
  color: PropTypes.string,
}
PhoneItem.defaultProps = {
  imageFileName: 'noimage.png',
  price: 0,
  onClick: () => {},
}

export { PhoneItem }
