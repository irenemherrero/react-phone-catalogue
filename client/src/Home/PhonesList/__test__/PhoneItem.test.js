import React from 'react'
import { PhoneItem } from '../PhoneItem'
import { render, cleanup, fireEvent } from '@testing-library/react'

const mockRequiredPhone = {
  id: 0,
  name: 'iPhone 7',
  color: 'black,blue',
}
const mockFullPhone = {
  ...mockRequiredPhone,
  manufacturer: 'Apple',
  description:
    'iPhone 7 dramatically improves the most important aspects of the iPhone experience. It introduces advanced new camera systems. The best performance and battery life ever in an iPhone. Immersive stereo speakers. The brightest, most colorful iPhone display. Splash and water resistance*. And it looks every bit as powerful as it is. This is iPhone 7.',
  price: 769,
  imageFileName: 'IPhone_7.png',
  screen: '4,7 inch IPS',
  processor: 'A10 Fusion',
  ram: 2,
}

describe('PhoneItem (a child of PhonesList)', () => {
  afterEach(cleanup)

  it('render correctly', () => {
    const component = render(<PhoneItem {...mockFullPhone} />)
    expect(component).toMatchSnapshot()
  })
  it('check default values', () => {
    const { getByText } = render(<PhoneItem {...mockRequiredPhone} />)
    const name = getByText(mockRequiredPhone.name)
    const manufacturer = getByText(PhoneItem.defaultProps.manufacturer)
    const price = getByText(`${PhoneItem.defaultProps.price} â‚¬`)

    expect(name).toBeTruthy()
    expect(manufacturer).toBeTruthy()
    expect(price).toBeTruthy()
  })
  it('click on phone', () => {
    const clickMock = jest.fn()
    const { container } = render(
      <PhoneItem {...mockFullPhone} onClick={clickMock} />
    )
    fireEvent.click(container.firstChild)
    expect(clickMock).toHaveBeenCalled()
  })
})
