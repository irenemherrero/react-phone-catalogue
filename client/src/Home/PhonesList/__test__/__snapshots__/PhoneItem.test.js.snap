// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneItem (a child of PhonesList) render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-fzoLsD hpPURA"
        data-testid="phone-item"
      >
        <div
          class="sc-fzpans gnBuKW"
        >
          <img
            alt="iPhone 7"
            class="sc-Axmtr cFtcXX object-cover"
            src="undefinedIPhone_7.png"
          />
          <section
            class="sc-fzplWN DAXdT"
          >
            <h1>
              iPhone 7
            </h1>
            <h2>
              Apple
            </h2>
            <div
              class="sc-AxhCb JKePI"
            >
              <span
                class="sc-AxhUy eyJlmj"
                color="black"
              />
              <span
                class="sc-AxhUy deRsvS"
                color="blue"
              />
            </div>
            <footer>
              769 €
            </footer>
          </section>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-fzoLsD hpPURA"
      data-testid="phone-item"
    >
      <div
        class="sc-fzpans gnBuKW"
      >
        <img
          alt="iPhone 7"
          class="sc-Axmtr cFtcXX object-cover"
          src="undefinedIPhone_7.png"
        />
        <section
          class="sc-fzplWN DAXdT"
        >
          <h1>
            iPhone 7
          </h1>
          <h2>
            Apple
          </h2>
          <div
            class="sc-AxhCb JKePI"
          >
            <span
              class="sc-AxhUy eyJlmj"
              color="black"
            />
            <span
              class="sc-AxhUy deRsvS"
              color="blue"
            />
          </div>
          <footer>
            769 €
          </footer>
        </section>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
